src/components/validation/validators/component_spec/sources.rs

```diff
@@ -0,0 +1,396 @@
+use std::fmt::{Display, Formatter};
+
+use bytes::BytesMut;
+use vector_core::event::{Event, MetricKind};
+use vector_core::EstimatedJsonEncodedSizeOf;
+
+use crate::components::validation::{encode_test_event, TestEvent, ValidationConfiguration};
+
+use super::filter_events_by_metric_and_component;
+
+const TEST_SOURCE_NAME: &str = "test_source";
+
+pub enum SourceMetrics {
+    EventsReceived,
+    EventsReceivedBytes,
+    ReceivedBytesTotal,
+    SentEventsTotal,
+    SentEventBytesTotal,
+}
+
+impl SourceMetrics {
+    const fn name(&self) -> &'static str {
+        match self {
+            SourceMetrics::EventsReceived => "component_received_events_total",
+            SourceMetrics::EventsReceivedBytes => "component_received_event_bytes_total",
+            SourceMetrics::ReceivedBytesTotal => "component_received_bytes_total",
+            SourceMetrics::SentEventsTotal => "component_sent_events_total",
+            SourceMetrics::SentEventBytesTotal => "component_sent_event_bytes_total",
+        }
+    }
+
+    pub const fn validation(
+        &self,
+    ) -> fn(
+        &ValidationConfiguration,
+        &[TestEvent],
+        &[Event],
+        &[Event],
+    ) -> Result<Vec<String>, Vec<String>> {
+        match self {
+            SourceMetrics::EventsReceived => validate_component_received_events_total,
+            SourceMetrics::EventsReceivedBytes => validate_component_received_event_bytes_total,
+            SourceMetrics::ReceivedBytesTotal => validate_component_received_bytes_total,
+            SourceMetrics::SentEventsTotal => validate_component_sent_events_total,
+            SourceMetrics::SentEventBytesTotal => validate_component_sent_event_bytes_total,
+        }
+    }
+
+    pub const fn all() -> [SourceMetrics; 5] {
```

Comment:
Non-blocking nit: you'd be able to drop the `.iter()` call on line 112 in `validators/component_spec/mod.rs` if you just made this return `&[SourceMetrics]` instead of a fixed-size array.

Should be as easy as slapping a `&` at the front of the array declaration... maybe an `[..]` at the end? One or both of those things. Also just a bit more idiomatic to return a slice unless the size of the array is somehow relevant to the caller.

Reply:
I dropped this code, actually.